### PR TITLE
update tokens_transfer.circom to include both pubkey_x and pubkey_y

### PR DESCRIPTION
**Pull request by therealyingtong**

### CODE REVIEW

**Comment by barryWhiteHat** [comment]
Looks good. I think we should be careful about the json files that we commit to version contorling. For example `package-lock.json` should definetly be excluded. If a file i automatically compiled then we should not version control it. Most of the other json files look like test vectors it seems reasonable to include them for testing. I will defer to you on that decision.

File: utils/update.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+const txLeaf = require("./generate_tx_leaf.js");
+const babyJub = require("../circomlib/src/babyjub");
+const snarkjs = require("snarkjs");
+const bigInt = snarkjs.bigInt;
+const createBlakeHash = require("blake-hash");
+
+module.exports = {
+
+    processTx: function(tx, fromLeaf, toLeaf, signature){
+        const sBuff = eddsa.pruneBuffer(createBlakeHash("blake512").update(prv).digest().slice(0,32));
```
Comment (barryWhiteHat) [inline utils/update.js]: We might want to use a hash function that is available in the EVM here as we may need to validate signatures in the EVM at some stage.

Comment (barryWhiteHat) [reply utils/update.js]: Probably not in scope to change it in this PR just a note.

Comment (therealyingtong) [reply utils/update.js]: sorry -- removed this from the utils, should have been using verifyMiMC which doesn't require Blake hash.

File: withdraw/tokens_transfer_and_withdraw.circom [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    signal output out;
+
+    var i;
+
+    var NONCE_MAX_VALUE = 100;
+
+    // constant zero address (EdDSA public key derived from "0x00000..." private key)
+                         
+    var ZERO_ADDRESS_X = 00000000000000000000000000000000000000000000000000000000000000000000000000000;
+    var ZERO_ADDRESS_Y = 0000000000000000000000000000000000000000000000000000000000000000000000000000;
```
Comment (therealyingtong) [inline withdraw/tokens_transfer_and_withdraw.circom]: uses 0x000... for zero address!

Comment (barryWhiteHat) [reply withdraw/tokens_transfer_and_withdraw.circom]: Cool. So I need to think about this more. Am thinking we should use a point on the jubjub curve 0, 0 is not and that might break the point addition formula. Need to think on this.

**Comment by barryWhiteHat** [comment]
This looks good. @therealyingtong will i merge?

**Comment by vaibhavchellani** [comment]
Would be great if we could have shorter filenames 😄

**Comment by therealyingtong** [comment]
yes sure! i tried not to touch `tokens_transfer` as far as possible, so most changes are in the `withdraw` and `multiple_tokens_transfer_and_withdraw`* folders

* i see what you mean @vaibhavchellani

**Comment by barryWhiteHat** [comment]
Why are you trying not to touch tokens_transfer? I think it would be good to include these changes there if possible.

**Comment by barryWhiteHat** [comment]
Oh wait i remember we said that its good not to touch there because multiple people were working on it. But lets check in the call to see how best to move forward.

File: tokens_transfer/tokens_transfer.circom [hunk 3]
```diff
@@ -118,11 +123,12 @@ template Main(n,m) {
...
     // accounts updates
-    component new_hash_from = MultiMiMC7(4,91);
-    new_hash_from.in[0] <== pubkey_x;
-    new_hash_from.in[1] <== token_balance_from-amount;
-    new_hash_from.in[2] <== nonce_from+1;
-    new_hash_from.in[3] <== token_type_from;
+    component new_hash_from = MultiMiMC7(5,91);
+    new_hash_from.in[0] <== from_x;
+    new_hash_from.in[1] <== from_y;
+    new_hash_from.in[2] <== token_balance_from-amount;
```
Comment (fredlacs) [inline tokens_transfer/tokens_transfer.circom]: This is the refactoring I mentioned during the call `token_balance_from - amount`

Comment (fredlacs) [reply tokens_transfer/tokens_transfer.circom]: but from reading the conversation I see you tried not to touch the token_transfers hah

**Comment by fredlacs** [comment]
Should we try and merge it this week? @barryWhiteHat @therealyingtong

**Comment by barryWhiteHat** [comment]
I kinda want to merge to token_transfer and have that be our only snark circuit for the time being. But we are still trying to figure the bset work flow out. 

Are you blocked by this?

**Comment by fredlacs** [comment]
> I kinda want to merge to token_transfer and have that be our only snark circuit for the time being. But we are still trying to figure the bset work flow out.
> 
> Are you blocked by this?

All is good, copied in the utils manually in the meantime

**Pull request merged**